src/cache.rs

```diff
@@ -637,15 +637,23 @@ impl Cache {
     pub fn resolve_imports(
         &mut self,
         file_id: FileId,
-    ) -> Result<CacheOp<()>, CacheError<ImportError>> {
+    ) -> Result<CacheOp<Vec<FileId>>, CacheError<ImportError>> {
         match self.entry_state(file_id) {
-            Some(state) if state >= EntryState::ImportsResolved => Ok(CacheOp::Cached(())),
+            Some(state) if state >= EntryState::ImportsResolved => Ok(CacheOp::Cached(Vec::new())),
             Some(state) if state >= EntryState::Parsed => {
-                if state < EntryState::ImportsResolving {
+                let pending = if state < EntryState::ImportsResolving {
                     let CachedTerm {
                         term, parse_errs, ..
-                    } = self.terms.remove(&file_id).unwrap();
+                    } = self.terms.get(&file_id).unwrap();
+                    // okay, for now these clones are here becuase the function call below
+                    // short circuts, and then we don't put back the item in cache, so the
```

Comment:
I'm not sure to fully understand this comment. What do you mean by "short-circuits"?

Reply:
And what is the "function below"? Probably better to name it explicitly

Reply:
So previously, before we resolve an import term in a file, we (1) remove the file's `RichTerm` from the cache, (2) resolve the import, (3) update the file's `RichTerm` and (4) put it back in the cache. However, the "resolve the import" phase can fail for whatever reason (e.g wrong file path, maybe parse errors, etc), and if it fails, we no longer put the `RichTerm` back in the cache, because of the monadic operator `?` (this is what I mean by short circuits), and the file cannot be linearized if the `RichTerm` is not present, so the LSP crashes.

Reply:
What about putting it back in the case `resolve_imports` return an error? I'm just curious though, since cloning a term isn't very expensive, I guess it's fine and less error-prone this way.

Reply:
The initial term is moved by the call to `import_resolution::resolve_imports(term, self)`, and if it fails, there's nothing to put back in the cache

Reply:
Ah, right. Then cloning it is :shrug: